mineinabyss-features/src/main/kotlin/com/mineinabyss/armorstandlock/ArmorStandLockingListener.kt

```diff
@@ -92,4 +99,114 @@ class ArmorStandLockingListener : Listener {
+
+        if (caught == null) return // We only care about non-null caught entities
+        if (caught!!.type != EntityType.ARMOR_STAND) return // We only care about armor stands
+
+        val armorStandEntity = caught as ArmorStand
+        val armorStand = armorStandEntity.toGeary().get<LockArmorStand>() ?: return
+
+        if (!armorStand.lockState) return
+
+        // The player is allowed to fish if they are the owner of the armor stand or if they have the permission to bypass the lock
+
+        if (!armorStand.isAllowed(player.uniqueId) && !player.hasPermission("mineinabyss.lockarmorstand.bypass")) {
+            player.error("You do not have access to fishing this armor stand!")
+            hook.remove()
+            isCancelled = true
+        }
+    }
+
+    private fun onPistonExtendRetract(block: Block, direction: BlockFace, blocks: List<Block>): Boolean {
+        /**
+         * This function is used to prevent pistons from extending armor stands,
+         * to prevent pistons from pushing blocks into armor stands,
+         * and to prevent pistons from pushing armor stands using honey blocks below them.
+         */
+
+        val blocksThePistonChanges = mutableListOf<Block>()
+
+        blocksThePistonChanges.add(block.location.add(direction.direction).block)
+
+        for (blockinblocks in blocks) {
+            // Gather all blocks that the piston will probably change in a list
+            blocksThePistonChanges.add(blockinblocks.location.add(direction.direction).block)
+
+            if (blockinblocks.type == Material.HONEY_BLOCK) {
+                // If the piston is pushing honey blocks, we need to check if the honey block is below the armor stand
+                blocksThePistonChanges.add(blockinblocks.location.add(BlockFace.UP.direction).block)
+            }
+        }
+
+        for (blockinblocks in blocksThePistonChanges) {
+            // Get all armor stands in the blocks the piston changes
+            val armorStands = blockinblocks.location.getNearbyEntitiesByType(ArmorStand::class.java, 1.0)
+            for (armorStand in armorStands) {
+                // If any armorstand is locked, cancel the event
+                val gearyArmorStand = armorStand.toGeary().get<LockArmorStand>() ?: continue
+                if (!gearyArmorStand.lockState) continue
+
+                return true
+            }
+        }
+
+        // Everything is fine, the piston can move :)
+
+        return false
+    }
+
+    @EventHandler
+    fun BlockPistonExtendEvent.onPistonExtend() {
+        /**
+         * This listener is used to prevent pistons from pushing blocks into armor stands.
+         * It calls the onPistonExtendRetract function to check if the piston is pushing blocks into armor stands.
+         */
+        if (onPistonExtendRetract(block, direction, blocks)) {
+            isCancelled = true
+        }
```

Comment:
Perhaps `blocks.remove(block)` is better than cancelling the entire event?

Reply:
The blocks list is immutable, this isn't allowed by the event.